### PR TITLE
GODRIVER-2966 Remove the heartbeat events after a connection is closed / a check is canceled.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-2966

## Summary
Remove the heartbeat events when checking on the server after a connection is closed / a check is canceled.

### CODE REVIEW

**Review by prestonvasquez — DISMISSED** [review]
LGTM! Did you notice any freezing with the lambda update? Which CI tasks would this be tested in?

**Comment by qingyang-hu** [comment]
Unfortunately, we don't have a CI task to test against the lambda freezing. I can try to add one.

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Comment by prestonvasquez** [comment]
@qingyang-hu Currently there is no clear requirement in the SDAM specifications that requires us to publish heartbeat events when attempting to establish a connection. However, many drivers do this including Python, Node, and Rust. 

Go Driver only publishes the "started" event when the connection is not nil. So we don't publish this event on the initial connection, per se. Only when we re-establish connections. The reason we do this comes from PR #1133 , where IIUC it was determined that passing an empty value for the connection ID is not appropriate. There is a Drivers ticket open for this right now, [ DRIVERS-2677](https://jira.mongodb.org/browse/DRIVERS-2677) , where it has been suggested that passing a "0" value for the connectionID is a viable solution in this case since connectionID starts at "1".

I suggest we do three things to resolve this PR: 

1. Remove the if block here: https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L801 , but leave the publication. When the connection ID is unknown, then publish that values as "0". 

2. Remove this entire block: https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L811-L813 . IIUC, it is a duplicate of this publication: https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L865 . Is that correct?

3. Remove this entire block: https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L816-L818 . IIUC, it is a duplicate of this publication: https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L872 . Is that correct?

**Comment by qingyang-hu** [comment]
I think, instead, we need to remove all publications in the `else` block [L820-L874](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L820-L874) including [L829](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L829) if we plan to publish heartbeat events when attempting to establish a connection.

**Comment by prestonvasquez** [comment]
> I think, instead, we need to remove all publications in the `else` block [L820-L874](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L820-L874) including [L829](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L829) if we plan to publish heartbeat events when attempting to establish a connection.

The publications in the else block are required by the specification, i.e. before we send a hello command and whether or not the hello failed/succeeded: https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.rst#events-api

File: x/mongo/driver/topology/server.go [hunk 1]
```diff
@@ -795,12 +795,15 @@ func (s *Server) check() (description.Server, error) {
...
+	// check, or the previous check was cancelled.
 	if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
-		// Create a new connection if this is the first check, the connection was closed after an error during the previous
-		// check, or the previous check was cancelled.
+		connID := "0"
 		if s.conn != nil {
-			s.publishServerHeartbeatStartedEvent(s.conn.ID(), false)
+			connID = s.conn.ID()
 		}
+		s.publishServerHeartbeatStartedEvent(connID, false)
```
Comment (matthewdale) [inline x/mongo/driver/topology/server.go:806]: We should do something similar for the subsequent `publishServerHeartbeatSucceededEvent` and `publishServerHeartbeatFailedEvent` so they always get published, even if the conn is nil. That's especially important for the failed event so we can report the error.

**Review by prestonvasquez — APPROVED** [review]
LGTM

**Review by prestonvasquez — APPROVED** [review]

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Pull request merged**